kaldb/src/main/java/com/slack/kaldb/bulkIngestApi/opensearch/BulkApiRequestParser.java

```diff
@@ -112,14 +112,22 @@ public static Trace.Span fromIngestDocument(
     sourceAndMetadata.remove(IngestDocument.Metadata.ID.getFieldName());
     sourceAndMetadata.remove(IngestDocument.Metadata.INDEX.getFieldName());
 
-    sourceAndMetadata.forEach(
-        (key, value) -> spanBuilder.addTags(SpanFormatter.convertKVtoProto(key, value, schema)));
-    spanBuilder.addTags(
-        Trace.KeyValue.newBuilder()
-            .setKey(SERVICE_NAME_KEY)
-            .setVType(Trace.ValueType.STRING)
-            .setVStr(index)
-            .build());
+    boolean tagsContainServiceName = false;
+    for (Map.Entry<String, Object> kv : sourceAndMetadata.entrySet()) {
+      if (!tagsContainServiceName && kv.getKey().equals(SERVICE_NAME_KEY)) {
+        tagsContainServiceName = true;
+      }
+      spanBuilder.addTags(SpanFormatter.convertKVtoProto(kv.getKey(), kv.getValue(), schema));
+    }
+    if (!tagsContainServiceName) {
```

Comment:
making sure we don't add `service_name` twice as a tag. We will anyways pick the first entry in the indexer, but this is for lesser ambiguity in the future.

Reply:
A more readable method of doing this might be:

```java
    sourceAndMetadata.forEach((key, value) -> spanBuilder.addTags(SpanFormatter.convertKVtoProto(key, value, schema)));
    if (sourceAndMetadata.entrySet().stream().noneMatch((entry) -> entry.getKey().equals(SERVICE_NAME_KEY))) {
      spanBuilder.addTags(
          Trace.KeyValue.newBuilder()
              .setKey(SERVICE_NAME_KEY)
              .setVType(Trace.ValueType.STRING)
              .setFieldType(Schema.SchemaFieldType.KEYWORD)
              .setVStr(index)
              .build());
    }
```

Reply:
Note that in the above example it would require traversing the tags up to twice, which may not be desired since this is a hot path.